### PR TITLE
Fix collection modified bug in AssignPendingVectors 

### PR DESCRIPTION
**Pull request by Derrick-**
Fix collection modified bug in AssignPendingVectors
Code cleanup in separate commits.

Fixes: https://github.com/stratisproject/StratisBitcoinFullNode/issues/126

### CODE REVIEW

**Comment by dangershony** [comment]
Argh thats hard to review.

File: Stratis.Bitcoin/BlockPulling/BlockPuller.cs [hunk 1]
```diff
@@ -267,7 +267,7 @@ private void AssignPendingVectors()
 			}
 
 			var minheight = int.MaxValue;
-			foreach (InventoryVector vector in vectors)
+			foreach (InventoryVector vector in vectors.ToArray())
```
Comment (Derrick-) [inline Stratis.Bitcoin/BlockPulling/BlockPuller.cs]: This is the only functional change. Rest is refactoring.
See individual commits.

File: Stratis.Bitcoin/BlockPulling/BlockPuller.cs [hunk 2]
```diff
@@ -22,14 +36,21 @@ public class DownloadedBlock
 
 		public class BlockPullerBehavior : NodeBehavior 
 		{
+
+			private readonly CancellationTokenSource cancellationToken = new CancellationTokenSource();
```
Comment (dangershony) [inline Stratis.Bitcoin/BlockPulling/BlockPuller.cs]: I think we should not instantiate params on creation but in the constructor, it makes it really hard to debug that way.

File: Stratis.Bitcoin/BlockPulling/BlockPuller.cs (hunk 1 above)
Comment (dangershony) [inline Stratis.Bitcoin/BlockPulling/BlockPuller.cs]: This change is great

File: Stratis.Bitcoin/BlockPulling/BlockPuller.cs (hunk 2 above)
Comment (Derrick-) [inline Stratis.Bitcoin/BlockPulling/BlockPuller.cs]: Ok. Will change.

**Comment by dangershony** [comment]
Did you get a chance to run the tests?

**Pull request merged**